crates/shared/src/price_estimation/competition.rs

```diff
@@ -79,33 +79,50 @@ impl<T: Send + Sync + 'static> RacingCompetitionEstimator<T> {
     ) -> futures::future::BoxFuture<'_, Result<R, E>> {
         async move {
             let mut results = vec![];
+            let mut iter = self.inner.iter().enumerate().peekable();
             // Process stages sequentially
-            'outer: for (stage_index, stage) in self.inner.iter().enumerate() {
-                // Process estimators within each stage in parallel
-                let mut futures: Vec<_> = stage
-                    .iter()
+            'outer: while let Some((stage_index, stage)) = iter.next() {
+                // Make sure we also use the next stage(s) if this one does not have enough
+                // estimators to return early anyways
+                let mut requests = stage
+                    .into_iter()
                     .enumerate()
-                    .map(|(index, (_, estimator))| {
+                    .map(|(index, (name, estimator))| {
+                        (EstimatorIndex(stage_index, index), name, estimator)
+                    })
+                    .collect::<Vec<_>>();
+                while stage.len() < self.successful_results_for_early_return.get()
+                    && iter.peek().is_some()
```

Comment:
This could cause issues for a fully pipelined price estimation strategy (all stages with a single estimator). In that case you'd activate all stages in parallel even if you only need 2.

I think what you want to do is sth like `while number_successful_responses + prepared_requests + stage.len() < threshold`. That way you always query exactly as many stages in parallel as you need to get over the threshold given the successful responses you already have assuming all responses of this batch are `Ok`.

Reply:
Ah, very good catch. Originally I extended stages, but now it's requests. Fixed and adjusted the unit test to catch this.

Reply:
The new version is still too aggressive with batching requests IMO.
Lets say you have a config with 4 stages with 1 estimator each and a threshold of 2.
In the first batch you'll trigger requests for stage 1 and 2. Now let's assume there is 1 good result in there.
Then it would be enough to simply trigger stage 3 because the 1 missing request might put us over the threshold but currently we would trigger stage 3 and 4 together because each batch is supposed to be able to pass the threshold on its own.
Granted this is a bit contrived but but I think if we already introduce hard to read/understand logic for this optimization we might as well make it perfect.

Reply:
You are right. Should be fixed now. To make it really perfect we would start scheduling new estimators as some estimators of the current stage finish unsuccessfully (and there is no hope for the current stage to be sufficient by itself), but I cannot think of a way to achieve this without making this code even harder to read.